### PR TITLE
Allow setting sign-passphrase directly

### PR DESCRIPTION
**Pull request by reavessm**
Signed-off-by: Stephen Reaves <reaves735@gmail.com>

#### Does this PR introduce a user-facing change?

```release-note
Added the '--sign-passphrase' option to allow users to specify gpg signing passphrase directly.
```

### CODE REVIEW

**Comment by reavessm** [comment]
the CI is saying:

```
xref-helpmsgs-manpages: 'podman image push --help' lists '--sign-passphrase', which is not in docs/source/markdown/podman-push.1.md
xref-helpmsgs-manpages: 'podman push --help' lists '--sign-passphrase', which is not in docs/source/markdown/podman-push.1.md
```

But they're clearly [here](https://github.com/containers/podman/pull/15840/files#diff-c7b044b01a2f89b86d880dd61d7525ea6e017f9465eb58e6cac6f4e6f8f288e5R129-R131).  Can I get some help here?  Did I add them in the wrong spot?

**Comment by mheon** [comment]
@edsantiago PTAL

**Comment by edsantiago** [comment]
You committed a new file, `podman-push.1.md`. You can't do that. Please modify the `.in` file instead.

File: cmd/podman/manifest/push.go [hunk 1]
```diff
@@ -83,6 +83,10 @@ func init() {
 	flags.StringVar(&manifestPushOpts.SignPassphraseFileCLI, signPassphraseFileFlagName, "", "Read a passphrase for signing an image from `PATH`")
 	_ = pushCmd.RegisterFlagCompletionFunc(signPassphraseFileFlagName, completion.AutocompleteDefault)
 
+	signPassphraseFlagName := "sign-passphrase"
+	flags.StringVar(&manifestPushOpts.SignPassphrase, signPassphraseFlagName, "", "Passphrase to use for signing an image, overrides 'sign-passphrase-file'")
```
Comment (rhatdan) [inline cmd/podman/manifest/push.go:87]: This option should conflict with --sign-passphrase-file.
If user uses both then throw an error.

File: docs/source/markdown/podman-manifest-push.1.md.in [hunk 2]
```diff
@@ -55,6 +55,10 @@ Sign the pushed images with a “simple signing” signature using the specified
 
 Sign the pushed images with a sigstore signature using a private key at the specified path. (This option is not available with the remote Podman client, including Mac and Windows (excluding WSL2) machines)
 
+#### **--sign-passphrase**=*passphrase*
+
+If signing the image with **--sign-by**, set the signing passphrase
```
Comment (rhatdan) [inline docs/source/markdown/podman-manifest-push.1.md.in:60]: Add conflicts with --sign-passphraze-file.

And add conflicts with --sign-passphrase below

File: docs/source/markdown/podman-push.1.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
+% podman-push 1
```
Comment (rhatdan) [inline docs/source/markdown/podman-push.1.md:1]: Remove this file.

File: test/e2e/push_test.go [hunk 4]
```diff
@@ -166,6 +166,18 @@ var _ = Describe("Podman push", func() {
 				Expect(pull.ErrorToString()).To(ContainSubstring("A signature was required, but no signature exists"))
 
 				// Sign an image, and verify it is accepted.
+				pass, err := os.ReadFile("testdata/sigstore-key.key.pass")
+				Expect(err).ShouldNot(HaveOccurred())
+				push = podmanTest.Podman([]string{"push", "-q", "--tls-verify=false", "--remove-signatures", "--sign-by-sigstore-private-key", "testdata/sigstore-key.key", "--sign-passphrase", string(pass), ALPINE, "localhost:5000/sigstore-signed"})
```
Comment (rhatdan) [inline test/e2e/push_test.go:171]: This test should fail and throw an error about the conflict.

**Comment by rhatdan** [comment]
Thanks @reavessm 
Once this gets merged, could you open a Similar PR for `buildah push`

**Review by vrothberg — COMMENTED** [review]
Needs a head nod from @mtrmac.

I am not in favor of adding passphrases on the command line.  Other process can easily explore the CLI (and env) via `/proc`.  Using a file is much more secure.

**Comment by vrothberg** [comment]
> Thanks @reavessm Once this gets merged, could you open a Similar PR for `buildah push`

Let's take a step a back and discuss the security concerns first.  Opening an issue ahead of the PR would have helped ironing out design decisions.

**Comment by mtrmac** [comment]
Yeah, I’m not happy about making this easy and I’d prefer for this option not to be implemented.

Every user on the system (including “unprivileged” daemons) can see the command lines of other users, revealing the password. Even on a single-user system, that violates the principle of least privilege and exposes the data to way too many accidents. (Erasing in-process secret keys after use was not even protecting data against any recognizable privilege boundary, and was a purely theoretical hassle required by bureaucrats for bad reasons — until Heartbleed happened.)

Stubborn enough users can already submit passphrases without using temporary files, using e.g. process substitution; but this is not exposed to be _easy_ for a reason.

If anything, I’d recommend thinking about making it easy to be more secure, devising a SELinux context for passphrase files so that Podman-only passphrase files can only be read by Podman and not by `unconfined_t`.

**Comment by reavessm** [comment]
@mtrmac @vrothberg  That's a fair concern that I hadn't considered.  As I'm still fighting with these unit tests, I'd be ok with closing this PR.  FWIW though, this doesn't get rid of the more secure option, this only adds a more convenient option.

**Comment by mtrmac** [comment]
Almost all users, almost all the time, are going chose the most convenient option that’s readily available. So I think it’s important for the most convenient option not to be insecure.

There’s certainly a balance to be struck; at _some_ point users are going to actively do work, and share tips, on how to bypass the intended secure design to get a more convenient system. I don’t think we are at that point here — distributing or version-controlling a single script, vs. a script + a passphrase file, seems about equally difficult, but I might be missing some case that needs to be made easier to do securely.

**Comment by vrothberg** [comment]
I'd be more comfortable with not exposing this option as well.  I hope this does not discourage you from further contributing to the project, @reavessm!

**Comment by rhatdan** [comment]
Would --sign-passphrase-file /dev/stdin work?

Then could you 

echo mysecret | podman push --sign-passphrease-file /dev/stdin ...

We could simplify this to allow

echo mysecret | podman push --sign-passphrease-file - ...

Then translate the `-` to /dev/stdin.

**Comment by mtrmac** [comment]
It might be useful to discuss the use case — how is signing being used so that including the passphrase on the command line seems to be the most appropriate option? Is it just some script where it seems to be easier to maintain a single file instead of two? Or something quite different we didn't expect?

**Comment by vrothberg** [comment]
Based on the conversation I am closing the PR.  Thanks a lot for proposing the change!

**Pull request closed without merging**